### PR TITLE
Optimization

### PR DESCRIPTION
**Pull request by my4ng**
Benchmark result (CK3-tiger):
- 0-2% for small mods
- ~15% for large mods

### CODE REVIEW

File: src/ck3/data/provinces.rs [hunk 1]
```diff
@@ -15,10 +16,43 @@ use crate::token::{Loc, Token};
 
 pub type ProvId = u32;
 
-#[derive(Clone, Debug, Default)]
+const COLOUR_COUNT: usize = 256 * 256 * 256;
+#[derive(Clone, Copy, Debug, Default)]
```
Comment (amtep) [inline src/ck3/data/provinces.rs:20]: Should it really be Copy? This is a 2MB array, right?

File: src/datatype.rs [hunk 2]
```diff
@@ -749,201 +752,140 @@ fn lookup_global_promote(lookup_name: &str) -> Option<(Args, Datatype)> {
...
-    lookup_name: &str,
-    ltype: Datatype,
-    global: &[(&str, Datatype, Args, Datatype)],
-) -> LookupResult {
-    let start = global.partition_point(|(name, _, _, _)| name < &lookup_name);
-    let mut found_any = false;
+#[inline]
+fn lookup_global_function(lookup_name: &str) -> Option<(Args, Datatype)> {
+    global_function(lookup_name)
+}
```
Comment (amtep) [inline src/datatype.rs:770]: Why these tiny functions? I don't understand why `lookup_global_function` and `global_function` are separate

File: src/datatype.rs [hunk 3]
```diff
@@ -749,201 +752,140 @@ fn lookup_global_promote(lookup_name: &str) -> Option<(Args, Datatype)> {
...
+    fn eq(&self, other: &Self) -> bool {
+        self.0.eq_ignore_ascii_case(other.0)
+    }
+}
+
+impl Eq for CaseInsensitiveStr {}
+
+impl std::hash::Hash for CaseInsensitiveStr {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.0.to_ascii_lowercase().hash(state);
```
Comment (amtep) [inline src/datatype.rs:848]: This might be an expensive hash since it allocates a String. How often is it called?

**Review by amtep — COMMENTED** [review]
Just two questions :)

File: src/datatype.rs (hunk 2 above)
Comment (my4ng) [reply src/datatype.rs:770]: This is purely to mirror promotes and functions, but yeah i think they should be removed

File: src/datatype.rs (hunk 3 above)
Comment (my4ng) [reply src/datatype.rs:848]: This is called for every insertion and every query. I don't think there is anyway to avoid this unfortunately.

**Review by amtep — APPROVED** [review]

**Pull request merged**